crates/router/src/core/routing/helpers.rs

```diff
@@ -543,3 +551,332 @@ pub fn get_default_config_key(
+                .to_string(),
+        })?
+        .algorithm_id
+        .ok_or(errors::ApiErrorResponse::GenericNotFoundError {
+            message: "unable to find algorithm id in success based algorithm config".to_string(),
+        })?;
+
+    let key = format!(
+        "{}_{}",
+        business_profile.get_id().get_string_repr(),
+        success_based_routing_id.get_string_repr()
+    );
+
+    if let Some(config) =
+        get_cached_success_based_routing_config_for_profile(state, key.as_str()).await
+    {
+        Ok(config.as_ref().clone())
+    } else {
+        let success_rate_algorithm = state
+            .store
+            .find_routing_algorithm_by_profile_id_algorithm_id(
+                business_profile.get_id(),
+                &success_based_routing_id,
+            )
+            .await
+            .change_context(errors::ApiErrorResponse::ResourceIdNotFound)
+            .attach_printable("unable to retrieve success_rate_algorithm for profile from db")?;
+
+        let success_rate_config = success_rate_algorithm
+            .algorithm_data
+            .parse_value::<routing_types::SuccessBasedRoutingConfig>("SuccessBasedRoutingConfig")
+            .change_context(errors::ApiErrorResponse::InternalServerError)
+            .attach_printable("unable to parse success_based_routing_config struct")?;
+
+        refresh_success_based_routing_cache(state, key.as_str(), success_rate_config.clone()).await;
+
+        Ok(success_rate_config)
+    }
+}
+
+/// metrics for success based dynamic routing
+#[cfg(all(feature = "v1", feature = "dynamic_routing"))]
+#[instrument(skip_all)]
+pub async fn push_metrics_for_success_based_routing(
+    state: &SessionState,
+    payment_attempt: &storage::PaymentAttempt,
+    routable_connectors: Vec<routing_types::RoutableConnectorChoice>,
+    business_profile: &domain::Profile,
+) -> RouterResult<()> {
+    let client = state
+        .grpc_client
+        .dynamic_routing
+        .success_rate_client
+        .as_ref()
+        .ok_or(errors::ApiErrorResponse::GenericNotFoundError {
+            message: "success_rate gRPC client not found".to_string(),
+        })?;
+
+    let payment_connector = &payment_attempt.connector.clone().ok_or(
+        errors::ApiErrorResponse::GenericNotFoundError {
+            message: "unable to derive payment connector from payment attempt".to_string(),
+        },
+    )?;
+
+    let success_based_routing_configs =
+        fetch_success_based_routing_configs(state, business_profile)
+            .await
+            .change_context(errors::ApiErrorResponse::InternalServerError)
+            .attach_printable("unable to retrieve success_rate configs")?;
+
+    let tenant_business_profile_id = format!(
+        "{}:{}",
+        state.tenant.redis_key_prefix,
+        business_profile.get_id().get_string_repr()
+    );
+
+    let success_based_connectors = client
+        .calculate_success_rate(
```

Comment:
why is this function not accepting `&str`?

Reply:
cc: @Aprabhat19

Reply:
For id_type we are anyhow passing a String value to the service, so it did not make sense to pass a &str and then take the ownership of the value later